adoc/chapters/cuda_backend.adoc

```diff
@@ -0,0 +1,598 @@
+
+[[table.cuda.appinterop.nativeobjects]]
+.Types of native backend objects application interoperability
+[width="100%",options="header",cols="20%,20%,20%,40%"]
+|====
+| [code]#SyclType# | [code]#backend_input_t<backend::cuda, SyclType># | [code]#backend_return_t<backend::cuda, SyclType># | Description
+| [code]#platform# | `std::vector<CUdevice>`   | `std::vector<CUdevice>`  | A SYCL platform encapsulates a list of CUDA devices.
+| [code]#device#   | `CUdevice`                | `CUdevice`               | A SYCL device encapsulates a CUDA device.
+| [code]#context#  | `CUcontext`               | `std::vector<CUcontext>` | A SYCL context can encapsulate multiple CUDA contexts, however, it is not possible to create a SYCL context from multiple CUDA contexts.
+| [code]#queue#    | `CUstream`   | `CUstream` | A SYCL queue can encapsulates multiple CUDA stream, however, a SYCL queue can only be created from or produce one, and any synchronization required should be performed.
+| [code]#event#    | `CUevent`    | `CUevent`  | A SYCL event can encapsulates multiple CUDA events, however, a SYCL event can only be created from or produce one, and a CUevent produced from a SYCL event may or may not be valid, use `sycl::cuda::has_native_event` to query this.
+| [code]#buffer# | NA | `void *` | A SYCL buffer encapsulates a CUDA device pointer. If the SYCL buffer is a sub-buffer, the returned `void *` is offset to the beginning of the sub-buffer.
+|====
+
+[[table.cuda.appinterop.make_interop_APIs]]
+.[code]#make_*# Interoperability APIs for native backend objects.
+[width="100%",options="header",cols="40%,60%"]
+|====
+| CUDA interoperability function                                    |  Description
+| [code]#template<backend Backend> +
+platform + 
+make_platform(const backend_input_t<Backend, platform> &backendObject);# 
+        | Create a SYCL `platform` from a list of CUDA device, the list must contain at least one CUDA device. As the SYCL execution environment for the CUDA backend contains a fixed number of platforms that are enumerated via `sycl::platform::get_platforms()`. Calling this function does not create a new platform. Rather it merely creates a `sycl::platform` object that is a copy of one of the platforms from that enumeration.
+
+| [code]#template<backend Backend> +
+device +
+make_device(const backend_input_t<Backend, device> &backendObject);# 
+        | Construct a SYCL `device` from a CUDA device. As the SYCL execution environment for the CUDA backend contains a fixed number of devices that are enumerated via `sycl::device::get_devices()`. Calling this function does not create a new device. Rather it merely creates a `sycl::device` object that is a copy of one of the devices from that enumeration.
+
+| [code]#template<backend Backend> +
+context +
+make_context(const backend_input_t<Backend, context> &backendObject,
+                     const async_handler asyncHandler = {});# 
+        | Create a SYCL `context` from a CUDA context.
```

Comment:
@illuhad you make a good point, thanks for raising this. I think we could resolve this by clarifying the wording such that for some SYCL implementations the mapping of SYCL context to CUDA context is limited to the CUDA primary context, which would limit the usefulness of the feature when the implementation is based on the runtime API, however, would allow us to maintain a generic interface between the two implementation approaches.

So the wording would be clarified such that `get_native<context, backend::cuda>` is permitted to return the CUDA primary context and `make_context<backend::cuda>` is permitted to only accept the CUDA primary context, and throw an exception otherwise.

However, in order to allow users to reason about this there would need to be some kind of query which exposes whether the implementation supports interop with non-primary CUDA contexts. Aside from this particular use case, there is another situation where having such a query would be important. Within a host task, when interoperating with the native objects, it's important to know whether the SYCL implementation initializes non-primary contexts, in order to ensure that the native API calls within the host task are compatible.

I see a couple of ways we could do this:
* A. Introduce a query which tells the user specifically whether the SYCL implementation maps all SYCL contexts to the CUDA primary context or whether unique contexts are allocated, this provides the specific information we need in this case, but if perhaps less scalable.
* B. Introduce a broader query which tells the user whether the  SYCL implementation is using the driver API or the runtime API, this could be useful for other places where there is a difference between the drive API and the runtime API, though it may not reflect the case where an implementation combines the two APIs.

Reply:
Thanks for your thoughts @AerialMantis! I agree that what you are proposing are possible solutions. I have a hunch that in practice it might be easier to work with for users if they know if an implementation is runtime or driver API based rather than make the query more fine-grained and just about context interop.
E.g. if you know that your CUDA backend is runtime API-based, you can make other assumptions about backend interop. In that case, you can know that backend interop is generally just going to work without having to worry about contexts at all, which I think can be very valuable.
I wonder if it might be easier to understand and clearer if the backend specification just had two flavors, or profiles, so that we can expose the behavior of both approaches in a more obvious way.

I suspect that even if an implementation were to utilize both APIs, you could probably figure out a definition to classify them, most likely around how they behave regarding context interop, and whether the current CUDA device from `cudaSetDevice()` plays a role.
For example, hipSYCL already uses some driver API components for manual management of PTX modules, but still follows the interop behavior of the runtime API-approach.
Do you think there are use cases where such a classification is not possible?

Reply:
> I wonder if it might be easier to understand and clearer if the backend specification just had two flavors, or profiles, so that we can expose the behavior of both approaches in a more obvious way.

Just throwing this out there ... it is also an option to have two different CUDA backends.  This means:

* There would be two different enumerators in `enum backend`.
* Each backend would be responsible for defining its own interop specification (if it wanted to support interop).

Having a single backend interop specification is nice, but only if developers can reasonably write interop code that is portable across both CUDA backends.  If we think the two CUDA backends are just too different, it might be easier for users if each one had its own separate interop specification.

Reply:
@illuhad @gmlueck I think you make a good point, I am wary of introducing different profiles of the CUDA backend specification if possible, I think it would be valuable to have a single specification that all implementations can follow, though I recognize this may not be possible if there are cases we can't handle generically.

I've clarified the wording to state that both APIs are supported. The biggest concern there was regarding the context and whether the primary context is used. I've added a note that the CUcontexts that are returned from interop can include the primary context, and in fact, this is now what the DPC++ CUDA backend does. I suspect we may still need a query for whether the implementation is using the CUDA Driver API or CUDA Runtime API, but I'd like to avoid this if possible to avoid having any bifurcation of the specification.

> Also, I wonder: Do we believe that there will actually be a use case to invoke make_context with the same context obtained from get_native()? Or will having this query for runtime API backends just tempt users to do things that are unsupported?

This is a good point, I suspect we don't want to support this, it would like cause havoc with the ownership model.